File: src/dual_endpoint/3server_sign.ts
import { PrivateKey, PublicKey } from '@bsv/sdk/primitives';
import Transaction from '@bsv/sdk/transaction/Transaction';
import TransactionSignature from '@bsv/sdk/primitives/TransactionSignature';
import Script from '@bsv/sdk/script/Script';
import LockingScript from '@bsv/sdk/script/LockingScript';
import { createDualMultisigScript } from './1base_tx';

/**
 * 服务器端为双端费用池花费交易回签
 * 对应 Go 实现：SpendTXServerSign
 *
 * @param tx               需要签名的交易对象
 * @param targetAmount     输入的 satoshi 金额（即 multisig UTXO 的数额）
 * @param serverPrivateKey 服务器私钥
 * @param clientPublicKey  客户端公钥
 * @returns 服务器签名字节（DER + sighash flag）
 */
export function spendTXServerSign(
  tx: Transaction,
  targetAmount: number,
  serverPrivateKey: PrivateKey,
  clientPublicKey: PublicKey,
): number[] {
  const serverPublicKey = serverPrivateKey.toPublicKey();

  // 构造 2-of-2 优先级脚本
  const priorityScript: Script = createDualMultisigScript([serverPublicKey, clientPublicKey]);

  // 确保 sourceTransaction 信息存在
  if (!tx.inputs[0].sourceTransaction) {
    tx.inputs[0].sourceTransaction = new Transaction();
    tx.inputs[0].sourceTransaction.outputs = [];
  }

  // 设置源输出（index 固定为 0）
  const priorityLocking = new LockingScript();
  priorityLocking.chunks = priorityScript.chunks;
  tx.inputs[0].sourceTransaction.outputs[0] = {
    satoshis: targetAmount,
    lockingScript: priorityLocking,
  };

  // 生成 sighash 数据
  const sighashData = TransactionSignature.format({
    sourceTXID: tx.inputs[0].sourceTXID || '',
    sourceOutputIndex: tx.inputs[0].sourceOutputIndex,
    sourceSatoshis: targetAmount,
    transactionVersion: tx.version,
    otherInputs: [],
    outputs: tx.outputs,
    inputIndex: 0,
    subscript: priorityScript,
    inputSequence: tx.inputs[0].sequence || 1,
    lockTime: tx.lockTime,
    scope: TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID,
  });

  // 服务器签名
  const signature = serverPrivateKey.sign(sighashData);
  const signatureDER = signature.toDER() as number[];
  const serverSignBytes = [
    ...signatureDER,
    TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID,
  ];

  return serverSignBytes;
}

File: pkg/dual_endpoint/3server_sign.go
package chain_utils

import (
	"fmt"

	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	// primitives "github.com/bsv-blockchain/go-sdk/primitives/ec"
	multisig "github.com/spycat55/KeymasterMultisigPool/pkg/libs"

	tx "github.com/bsv-blockchain/go-sdk/transaction"
	sighash "github.com/bsv-blockchain/go-sdk/transaction/sighash"
)

// 服务器 回签
func SpendTXServerSign(
	transactionObject *tx.Transaction,
	targetAmount uint64,
	serverPrivateKey *ec.PrivateKey,
	clientPublicKey *ec.PublicKey,
) (*[]byte, error) {
	serverPublicKey := serverPrivateKey.PubKey()

	// 创建优先级脚本
	priorityScript, err := multisig.Lock([]*ec.PublicKey{serverPublicKey, clientPublicKey}, 2)
	if err != nil {
		return nil, fmt.Errorf("创建优先级脚本失败: %v", err)
	}

	// 设置输入的锁定脚本
	transactionObject.Inputs[0].SetSourceTxOutput(
		&tx.TransactionOutput{
			Satoshis:      targetAmount,
			LockingScript: priorityScript,
		},
	)

	sigHash := sighash.Flag(sighash.ForkID | sighash.All)
	aMultisigUnlockingScriptTemplate, err := multisig.Unlock([]*ec.PrivateKey{}, []*ec.PublicKey{serverPublicKey, clientPublicKey}, 2, &sigHash)
	if err != nil {
		return nil, fmt.Errorf("failed to create unlocking script template: %w", err)
	}

	// 重新签名所有输入
	serverSignByte, err := aMultisigUnlockingScriptTemplate.SignOne(transactionObject, 0, serverPrivateKey)
	if err != nil {
		return nil, fmt.Errorf("b 重新签名输入 %d 失败: %v", 1, err)
	}

	return serverSignByte, nil
}
